File: server/common/protocol.py
import struct
from common.bet import BetInfo

HEADER_SIZE = 4


def recv_message(sock):
    """Receives a length-prefixed message and returns one of:
    - None: EOF received, client is done sending bets
    - ["WINNERS", agency_id]: winner query for agency_id
    - list[BetInfo]: a batch of bets
    """
    header = recv_all(sock, HEADER_SIZE)
    length = struct.unpack('!I', header)[0]
    payload = recv_all(sock, length).decode('utf-8')

    if payload == "EOF":
        return None

    if payload.startswith("WINNERS|"):
        agency_id = payload.split("|")[1]
        return ["WINNERS", agency_id]

    fields = payload.split('|')
    agency = fields[0]
    bet_fields = fields[1:]

    bets = []
    for i in range(0, len(bet_fields), 5):
        nombre, apellido, dni, nacimiento, numero = bet_fields[i:i+5]
        bets.append(BetInfo(agency, nombre, apellido, dni, nacimiento, numero))
    return bets


def send_confirmation(sock, message: str):
    """Sends a length-prefixed confirmation message. Uses send_all to avoid short-writes."""
    data = message.encode('utf-8')
    header = struct.pack('!I', len(data))
    send_all(sock, header)
    send_all(sock, data)


def send_winners(sock, dnis: list):
    """Sends a length-prefixed list of winner DNIs separated by '|'."""
    payload = "|".join(dnis).encode('utf-8')
    header = struct.pack('!I', len(payload))
    send_all(sock, header)
    send_all(sock, payload)


def send_all(sock, data: bytes):
    """Loops over sock.send until all bytes are sent, avoiding short-writes."""
    sent = 0
    while sent < len(data):
        n = sock.send(data[sent:])
        if n == 0:
            raise OSError("connection closed during send")
        sent += n


def recv_all(sock, length: int) -> bytes:
    """Loops over sock.recv until exactly length bytes are read, avoiding short-reads.
    A partial read is not an error — the remaining bytes will arrive in subsequent calls."""
    buf = b''
    while len(buf) < length:
        chunk = sock.recv(length - len(buf))
        if not chunk:
            raise OSError("connection closed during recv")
        buf += chunk
    return buf

File: client/common/protocol.go
package common

import (
	"encoding/binary"
	"fmt"
	"net"
	"strings"
)

const headerSize = 4
const maxChunkSize = 8 * 1024

// sends a batch of bets as a single length-prefixed message over conn: 4-byte big-endian payload size followed by the bet fields joined by '|'
// Sent in chunks of at most 8KB to avoid large writes
// Uses sendAll to avoid short-writes
func SendBatch(conn net.Conn, clientID string, bets []BetInfo) error {
	payload := encodeBets(clientID, bets)

	header := make([]byte, headerSize)
	binary.BigEndian.PutUint32(header, uint32(len(payload)))

	if err := sendAll(conn, header); err != nil {
		return fmt.Errorf("failed to send length header: %w", err)
	}
	if err := sendAll(conn, payload); err != nil {
		return fmt.Errorf("failed to send payload: %w", err)
	}
	return nil
}

// serializes all bets into a flat '|' separated payload
// The server reads clientID once (first field) then groups remaining fields by 5.
func encodeBets(clientID string, bets []BetInfo) []byte {
	fields := []string{clientID}
	for _, bet := range bets {
		fields = append(fields, bet.nombre, bet.apellido, bet.dni, bet.nacimiento, bet.numero)
	}
	return []byte(strings.Join(fields, "|"))
}

// SendEOF signals the server that the client has finished sending all batches.
func SendEOF(conn net.Conn) error {
	data := []byte("EOF")
	header := make([]byte, headerSize)
	binary.BigEndian.PutUint32(header, uint32(len(data)))
	if err := sendAll(conn, header); err != nil {
		return err
	}
	return sendAll(conn, data)
}

// reads a length-prefixed response from the server
// Uses recvAll to avoid short-reads
func RecvConfirmation(conn net.Conn) (string, error) {
	header := make([]byte, headerSize)
	if err := recvAll(conn, header); err != nil {
		return "", fmt.Errorf("failed to read length header: %w", err)
	}

	length := binary.BigEndian.Uint32(header)
	buf := make([]byte, length)
	if err := recvAll(conn, buf); err != nil {
		return "", fmt.Errorf("failed to read confirmation: %w", err)
	}
	return string(buf), nil
}

// sendAll loops over conn.Write in chunks of at most maxChunkSize, avoiding short-writes.
func sendAll(conn net.Conn, data []byte) error {
	sent := 0
	for sent < len(data) {
		end := sent + maxChunkSize
		if end > len(data) {
			end = len(data)
		}
		n, err := conn.Write(data[sent:end])
		if err != nil {
			return err
		}
		sent += n
	}
	return nil
}

// recvAll loops over conn.Read until buf is completely filled, avoiding short-reads.
func recvAll(conn net.Conn, buf []byte) error {
	received := 0
	for received < len(buf) {
		n, err := conn.Read(buf[received:])
		if err != nil {
			return err
		}
		received += n
	}
	return nil
}
